File: csharp_v2_beta/test/MicroServiceWorkshop.Tests/RapidsRivers/RiverTest.cs
/* 
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 */
 
using MicroServiceWorkshop.RapidsRivers;
using Newtonsoft.Json.Linq;
using NUnit.Framework;

namespace MicroServiceWorkshop.Tests.RapidsRivers
{
    // Ensures Packets are well-formed or catch construction errors
    [TestFixture]
    public class RiverTest
    {
        private const string SolutionString =
            "{\"need\":\"car_rental_offer\"," +
            "\"user_id\":456," +
            "\"solutions\":[" +
            "{\"offer\":\"15% discount\"}," +
            "{\"offer\":\"500 extra points\"}," +
            "{\"offer\":\"free upgrade\"}" +
            "]," +
            "\"frequent_renter\":\"\"," +
            "\"sample_float_key\":1.25," +
            "\"system.read_count\":2," +
            "\"contributing_services\":[]}";

        private const string MissingComma =
            "{\"frequent_renter\":\"\" \"read_count\":2}";

        private const string NeedKey = "need";
        private const string UserIdKey = "user_id";
        private const string SampleFloatKey = "sample_float_key";
        private const string KeyToBeAdded = "key_to_be_added";
        private const string EmptyArrayKey = "contributing_services";
        private const string InterestingKey = "frequent_renter";
        private const string SolutionsKey = "solutions";
        
        private TestRapidsConnection _rapidsConnection;
        private River _river;

        [SetUp]
        public void SetUp()
        {
            _rapidsConnection = new TestRapidsConnection();
            _river = new River(_rapidsConnection);
            _rapidsConnection.Register(_river);
        }

        [Test]
        public void ValidJson()
        {
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void InvalidJson()
        {
            _river.Register(new TestRiver((connection, problems) =>
            {
                Assert.True(problems.HasErrors());
                Assert.That(problems.ToString(), Does.Contain("Invalid JSON format"));
            }));
            _rapidsConnection.Process(MissingComma);
        }

        [Test]
        public void RequiredStringKeyExists()
        {
            _river.Require(NeedKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.AreEqual("car_rental_offer", (string)jsonPacket[NeedKey]);
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void RequiredIntegerKeyExists()
        {
            _river.Require(UserIdKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.AreEqual(456, (int)jsonPacket[UserIdKey]);
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void RequiredFloatKeyExists()
        {
            _river.Require(SampleFloatKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.AreEqual(1.25, (float)jsonPacket[SampleFloatKey]);
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void RequiredMultipleKeysExist()
        {
            _river.Require(NeedKey, UserIdKey, SampleFloatKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.AreEqual("car_rental_offer", (string)jsonPacket[NeedKey]);
                Assert.AreEqual(456, (int)jsonPacket[UserIdKey]);
                Assert.AreEqual(1.25, (float)jsonPacket[SampleFloatKey]);
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void MissingRequiredKeyDetected()
        {
            _river.Require("missing_key");
            _river.Register(new TestRiver((connection, problems) =>
            {
                Assert.True(problems.HasErrors());
                Assert.That(problems.ToString(), Does.Contain("missing_key"));
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void ForbiddenKeyMissing()
        {
            _river.Forbid("forbidden_key_1", "forbidden_key_2");
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.That(warnings.ToString(), Does.Contain("forbidden_key_1"));
                Assert.That(warnings.ToString(), Does.Contain("forbidden_key_2"));
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void EmptyStringPassesForbiddenValidation()
        {
            _river.Forbid(InterestingKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.That(warnings.ToString(), Does.Contain(InterestingKey));
            }));
            _rapidsConnection.Process(SolutionString);
        }

        [Test]
        public void EmptyArrayPassesForbiddenValidation()
        {
            _river.Forbid(EmptyArrayKey);
            _river.Register(new TestRiver((connection, jsonPacket, warnings) =>
            {
                Assert.False(warnings.HasErrors());
                Assert.That(warnings.ToString(), Does.Contain(EmptyArrayKey));
            }));
            _rapidsConnection.Process(SolutionString);
        }

        // Understands a mock RapidsConnection to allow tests to send messages
        private class TestRapidsConnection : RapidsConnection
        {
            public override void Publish(string message) { }  // Ignore for this test
            internal void Process(string message)
            {
                foreach (IMessageListener l in Listeners) l.HandleMessage(this, message);
            }
        }

        // Understands a mock River that invokes different delegates on success or failure
        delegate void Packet(RapidsConnection connection, JObject jsonPacket, PacketProblems warnings);

        delegate void OnError(RapidsConnection connection, PacketProblems errors);

        private class TestRiver : River.IPacketListener
        {
            private readonly Packet _successDelegate;
            private readonly OnError _failureDelegate;

            public TestRiver(Packet successDelegate) : this(successDelegate, UnexpectedFailure) { }
            public TestRiver(OnError failureDelegate) : this(UnexpectedSuccess, failureDelegate) { }

            private TestRiver(Packet successDelegate, OnError failureDelegate)
            {
                _successDelegate = successDelegate;
                _failureDelegate = failureDelegate;
            }

            public void Packet(RapidsConnection connection, JObject jsonPacket, PacketProblems warnings)
            {
                _successDelegate( connection, jsonPacket, warnings);
            }

            public void OnError(RapidsConnection connection, PacketProblems errors)
            {
                _failureDelegate(connection, errors);
            }

            private static void UnexpectedSuccess(RapidsConnection connection, JObject jsonPacket,
                PacketProblems warnings)
            {
                Assert.Fail("Unexpected successDelegate parsing JSON packet. Packet is:\n"
                            + jsonPacket
                            + "\nConclusions from parsing/validation are:\n"
                            + warnings);
            }

            private static void UnexpectedFailure(RapidsConnection connection, PacketProblems errors)
            {
                Assert.Fail("Unexpected JSON packet problem(s):\n" + errors.ToString());
            }
        }
    }
}

File: ruby_v2/test/rapids_rivers/river_test.rb
# Copyright (c) 2017 by Fred George.
# May be used freely except for training; license required for training.

require 'pry'

require_relative "../test_helper"

require_relative "../../lib/rapids_rivers/river"

class RiverTest < MiniTest::Test
  SOLUTION_STRING =
      "{\"need\":\"car_rental_offer\"," +
        "\"user_id\":456," +
        "\"solutions\":[" +
        "{\"offer\":\"15% discount\"}," +
        "{\"offer\":\"500 extra points\"}," +
        "{\"offer\":\"free upgrade\"}" +
        "]," +
        "\"frequent_renter\":\"\"," +
        "\"system.read_count\":2," +
        "\"contributing_services\":[]}";

  def setup
    @rapids_connection = TestRapids.new
    @river = River.new(@rapids_connection)
    @service = TestService.new(self)
    @river.register(@service)
  end

  def test_json_valid
    @service.define_singleton_method :packet do |send_port, packet, warnings|
      refute_messages warnings
    end
    @rapids_connection.received_message SOLUTION_STRING
  end

  def test_json_invalid
    @service.define_singleton_method :on_error do |send_port, errors|
      assert_errors errors
    end
    @rapids_connection.received_message "{\"key\":value}"
  end

  def test_required_field
    @river.require 'need', 'user_id'
    @service.define_singleton_method :packet do |send_port, packet, warnings|
      refute_messages warnings
      packet.need = packet.need + "_extra"
      packet.user_id = packet.user_id + 14
    end
    @rapids_connection.received_message SOLUTION_STRING
  end

  def test_required_field_missing
    @river.require 'need', 'missing_key'
    @service.define_singleton_method :on_error do |send_port, errors|
      assert_errors errors
    end
    @rapids_connection.received_message SOLUTION_STRING
  end

  def test_forbidden_field
    @river.forbid 'frequent_renter', 'contributing_services'
    @service.define_singleton_method :packet do |send_port, packet, warnings|
      refute_messages warnings
      packet.frequent_renter = 'platinum'
      packet.contributing_services << 'a testing service'
    end
    @rapids_connection.received_message SOLUTION_STRING
  end

  def test_forbidden_field_exists
    @river.forbid 'frequent_renter', 'user_id'
    @service.define_singleton_method :on_error do |send_port, errors|
      assert_errors errors
    end
    @rapids_connection.received_message SOLUTION_STRING
  end

  private

    class TestRapids
      include RapidsConnection
    end

    class TestService

      def initialize test_instance
        @test = test_instance
      end

      def packet rapids_connection, packet, warnings
        throw "Unexpected invocation of Service::packet. Warnings were:\n#{warnings}"
      end

      def on_error rapids_connection, errors
        throw "Unexpected invocation of Service::on_error. Errors detected were:\n#{errors}"
      end

      private

        def refute_messages packet_problems
          @test.refute packet_problems.messages?, packet_problems.to_s
        end

        def assert_errors packet_problems
          @test.assert packet_problems.errors?, packet_problems.to_s
        end

    end

end
